Handle invalid FS type on mount.

govint opened the pull request:
Handle invalid FS type returned by a GET call to the plugin server and fail back to using ext2 as FS type in such a case. This handles both any corruption on the server side plus makes a best effort to mount a volume with a default FS type.

govint commented:
Fixes #638

pdhamdhere commented on esx_service/volume_kv.py:
```diff
@@ -82,7 +82,7 @@
 # This option is handled in the volume-plugin at the docker host, and tracked in volume metadata. 
 FILESYSTEM_TYPE = 'fstype'
-DEFAULT_FILESYSTEM_TYPE = ''
+DEFAULT_FILESYSTEM_TYPE = 'ext4'
```
Aren't old volumes formatted with ext2?

pdhamdhere commented on vmdk_plugin/plugin.go:
```diff
@@ -361,6 +365,14 @@ func (d *vmdkDriver) Mount(r volume.MountRequest) volume.Response {
+		log.WithFields(
+			log.Fields{"name": r.Name, "error": "Invalid filesystem type."},
+		).Error("Failed to mount ")
+		return volume.Response{Err: "Failed mount - invalid filesystem type."}
```
minor edit: "Failed to mount - Invalid filesystem type"

msterin commented on vmdk_plugin/plugin.go:
```diff
@@ -84,14 +84,18 @@ func getMountPoint(volName string) string {
 
 // Get info about a single volume
 func (d *vmdkDriver) Get(r volume.Request) volume.Response {
+	log.Debug("calling GET ")
```
I think this debug should be one layer above, when misc. methods are called.

msterin commented on vmdk_plugin/plugin.go:
```diff
@@ -84,14 +84,18 @@ func getMountPoint(volName string) string {
 func (d *vmdkDriver) Get(r volume.Request) volume.Response {
+	log.Debug("calling GET ")
 	status, err := d.ops.Get(r.Name)
+	log.Debug("returned GET ")
```
ditto

msterin commented on vmdk_plugin/plugin.go:
```diff
@@ -361,6 +365,14 @@ func (d *vmdkDriver) Mount(r volume.MountRequest) volume.Response {
+
+	if !exists {
+		log.WithFields(
+			log.Fields{"name": r.Name, "error": "Invalid filesystem type."},
```
can you use a var for message instead of copy-n-paste here and on line 374 ?  Also I believe the text ("failed mount - invalid filesystem type" ) should be the same in both cases, and ideally include some hint on what to do next. Though I have no idea what to do next ...

govint commented:
Updated changes, corrected error message as below:
  ---
$ docker run --rm -it -w /data -v dvol_2:/data busybox
docker: Error response from daemon: VolumeDriver.Mount: Mount error - got invalid filesystem type for dvol_2.
See 'docker run --help'.

And logs,
2016-10-23 21:14:58.662195329 -0700 PDT [INFO] Going into ServeUnix - Listening on Unix socket address="/run/docker/plugins/vmdk.sock"
2016-10-23 22:08:20.16849929 -0700 PDT [INFO] Mounting volume name="dvol_2"
2016-10-23 22:08:20.183811502 -0700 PDT [ERROR] name="dvol_2" error="Mount error - got invalid filesystem type for dvol_2"

govint commented:
With old volumes the client plugin just appears hung (no response causing
docker "run" to error, plugin responds to new commands but the volume used
for the mount is left with a refcount thats not going away till the client
is restarted. The behavior is documented in the description for the issue.

Yes, ext2 is the default.


On Fri, Oct 21, 2016 at 6:39 AM, Prashant Dhamdhere <
notifications@github.com> wrote:

> *@pdhamdhere* commented on this pull request.
>
> This is not an issue with new volumes.
> How are old volumes mounted with latest code? Can you please test using
> old volumes with new plugin and update Testing Done in PR?
> ------------------------------
>
> In esx_service/volume_kv.py
> <https://github.com/vmware/docker-volume-vsphere/pull/639#pullrequestreview-5185926>
> :
>
> > @@ -82,7 +82,7 @@
>  # Filesystem type
>  # This option is handled in the volume-plugin at the docker host, and tracked in volume metadata.
>  FILESYSTEM_TYPE = 'fstype'
> -DEFAULT_FILESYSTEM_TYPE = ''
> +DEFAULT_FILESYSTEM_TYPE = 'ext4'
>
> Aren't old volumes formatted with ext2?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/639#pullrequestreview-5185926>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APHseHSlGQUiDhBZMQgMFH0OO9wPWpE-ks5q2BCugaJpZM4KcERk>
> .
>
